weasyl/controllers/moderation.py

```diff
@@ -219,3 +219,192 @@ def modcontrol_copynotetostaffnotes_post_(request):
+        # Update the spam filtering backend to indicate that this was not spam
+        spam_filtering.submit(
+            is_ham=True,
+            user_ip=ip_addr,
+            user_agent_id=ua_id,
+            user_id=userid,
+            comment_type="journal",
+            comment_content=content,
+        )
+        welcome.journal_insert(userid=userid, journalid=journalid, rating=rating, settings=settings)
+    elif action == "reject":
+        moderation.hidejournal(journalid=journalid)
+    else:
+        raise HTTPSeeOther("/modcontrol/spamqueue/journal")
+
+    raise HTTPSeeOther("/modcontrol/spamqueue/journal")
+
+
+@moderator_only
+def modcontrol_spamqueue_submission_get_(request):
+    """
+    A Pyramid route controller for displaying any journal items which the spam filter has tagged as being spam.
+
+    Only displays items that have the `is_spam` flag set, and are not hidden (which is used as the 'reviewed' flag).
+
+    :param request: The Pyramid request.
+    :return: A rendered page for the journal spam queue.
+    """
+    query = define.engine.execute("""
+        SELECT lo.login_name, su.submitid, su.title, su.content
+        FROM login lo
+        INNER JOIN submission su USING (userid)
+        WHERE su.is_spam
+            AND su.settings !~ 'h'
+    """).fetchall()
+    return Response(define.webpage(
+        request.userid,
+        "modcontrol/spamqueue/submission.html",
+        [query],
+        title="Submission Spam Queue"
+    ))
+
+
+@moderator_only
+@token_checked
+def modcontrol_spamqueue_submission_post_(request):
+    """
+    A Pyramid route controller for approving for display or items which the spam filter has tagged as being spam.
+    :param request: The Pyramid request.
+    :return/raises: HTTPSeeOther for the journal spam queue.
+    """
+    action = request.params.get("directive")
+    submitid = request.params.get("submitid")
+
+    if not all([action, submitid]):
+        raise WeasylError("Unexpected")
+
+    if action == "approve":
+        # Approve and insert the journal into the notifications table.
+        submitid, userid, rating, settings, content, ua_id, ip_addr = define.engine.execute("""
+            UPDATE submission
+            SET is_spam = FALSE
+            WHERE submitid = %(id)s
+            RETURNING submitid, userid, rating, settings, content, submitter_user_agent_id, submitter_ip_address
+        """, id=submitid).first()
+        # Update the spam filtering backend to indicate that this was not spam
+        spam_filtering.submit(
+            is_ham=True,
+            user_ip=ip_addr,
+            user_agent_id=ua_id,
+            user_id=userid,
+            comment_type="submission",
+            comment_content=content,
+        )
+        welcome.submission_insert(userid=userid, submitid=submitid, rating=rating, settings=settings)
+    elif action == "reject":
+        # Note: We need to explicitly convert to int for the moment, otherwise this call fails with:
```

Comment:
This is more the rule than the exception – I wouldn’t expect `submitid` to be represented as a string in the other branch. So for now, how about:

```python
action = request.params['directive']
submitid = int(request.params['submitid'])
```

above, with no need for the `all([action, submitid])` check? At some point we can add something to get required values from `request.params` with a `WeasylError` otherwise, but compared to the usual user- and internet-facing endpoints, this one is less important because mods won’t mess with the parameters so any errors that reach Sentry are more likely to be alerting us to an actual mistake.